### ISSUE TITLE
Add support for AWS Redshift

### ISSUE THREAD

**Issue by sinabakh** [infracost/infracost#151]
Fixed by https://github.com/infracost/infracost/pull/587

**Comment by tim775** [comment]
I'm going to take a crack at this.

**Comment by alikhajeh1** [comment]
Thanks @tim775!

**Comment by tim775** [comment]
Looking at the [aws pricing page](https://aws.amazon.com/redshift/pricing/) and the terraform registry it seems like I can add pricing for [aws_redshift_cluster](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/redshift_cluster) resource.  There are also prices for 'spectrum', 'concurrency scaling', 'managed storage', and 'ML' but I don't see any way to map those to a terraform resource.

**Comment by alikhajeh1** [comment]
@tim775 `aws_redshift_cluster` would be great, "Backup storage" also seems hard to include given that it talks about anything above the cluster size is charged S3 rates (which we have as a supported resource).

I know it's too late for this ticket but we've started putting more details in tickets to make it easier for contributors, e.g. https://github.com/infracost/infracost/issues/574.

**Comment by aliscott** [comment]
Here's what I've found from a quick look:
* Spectrum - the AWS calculator allows this to be calculated by entering "Data scanned", so I think we can add a usage-based cost component with name `Spectrum data scanned` and key `monthly_spectrum_data_scanned_gb`.
* Concurrency scaling - from what I can tell this is enabled by default, but looks difficult to model. The AWS calculator doesn't include this so I would leave it out.
* Managed storage - Seems like this is charged if `node_type ~= ra3.*`.
* ML - doesn't seem to be supported by Terraform.

**Comment by tim775** [comment]
Ok I think that makes sense to me (for now anyway :)).
So the "main" cost component would look something like:
`
		{
			Name:           fmt.Sprintf("Redshift Cluster (%s)", nodeType),
			Unit:           "hours",
			UnitMultiplier: 1,
			HourlyQuantity: decimalPtr(decimal.NewFromInt(numberOfNodes)),
			ProductFilter: &schema.ProductFilter{
				VendorName:    strPtr("aws"),
				Region:        strPtr(region),
				...
				AttributeFilters: []*schema.AttributeFilter{
					{Key: "nodeType", Value: strPtr(nodeType)},
				},
			},
			PriceFilter: &schema.PriceFilter{
				PurchaseOption: strPtr("on_demand"),
			},
		},
`
then there would be an additional use cost component for spectrum like:
`
			{
				Name:            "Spectrum data scanned",
				Unit:            "terabytes scanned",
				UnitMultiplier:  1,
				MonthlyQuantity: <from u.Get("monthly_spectrum_data_scanned_gb")>
                                 ...
			}
`
then conditionaly another cost component for managed storage when nodeType is ra3.*
`
			{
				Name:            "Managed storage",
				Unit:            "GB-month",
				UnitMultiplier:  1,
				MonthlyQuantity: <from u.Get("monthly_storage_gb")> // I haven't looked for naming convetions yet
                                 ...
			}
`

**Comment by tim775** [comment]
So for "Concurrency scaling" it looks like the cost is equal to the hourly rate of the cluster billed at 1 second granularity (after you go over some limit).  Assuming there's no explicit price for concurrency scaling in the json, we _could_ make a usage cost component based on the cluster price:
```{
	Name:            "Excess concurrency scaling",
	Unit:            "seconds",
	UnitMultiplier:  1/3600,
	MonthlyQuantity: <from u.Get("monthly_excess_concurrency_scaling_seconds")>,
	ProductFilter: <same as nodeType>
	PriceFilter: <same as nodeType>
        ...
}```

But i'm not sure if deriving an price from another price is a sound thing to do.  Would that be ok or is the rule "If aws doesn't publish a price in the json, we don't put it in infracost"?

**Comment by aliscott** [comment]
> But i'm not sure if deriving an price from another price is a sound thing to do.  Would that be ok or is the rule "If aws doesn't publish a price in the json, we don't put it in infracost"?

We don't have a hard rule on this, not sure it has come up before. Just had a quick look at this and I see there are specific listed prices for the Concurrency Scaling in the Pricing API, using the following MongoDB query:

```
db.products.find({ vendorName: 'aws', service: 'AmazonRedshift', productFamily: 'Redshift Concurrency Scaling' }).pretty()
```

**Comment by alikhajeh1** [comment]
Released in v0.8.5
